lib/AST/GenericEnvironment.cpp

```diff
@@ -304,7 +304,23 @@ GenericEnvironment::getOrCreateArchetypeFromInterfaceType(Type depType) {
     }
   };
 
-  if (requirements.concreteType) {
+  auto canUseConcreteType = [&](Type concreteType) -> bool {
```

Comment:
I don't think this is valid because the existential signature here `<Self where Self == String>` drops the conformance requirement (it was redundant) so the opened archetype now has no constraints at all.

We could return an ErrorType here and diagnose such types somewhere, or just make it work, but right now this would have the wrong semantics.